internal/terraform/transform_check.go

```diff
@@ -0,0 +1,114 @@
+package terraform
+
+import (
+	"log"
+
+	"github.com/hashicorp/terraform/internal/addrs"
+	"github.com/hashicorp/terraform/internal/configs"
+	"github.com/hashicorp/terraform/internal/dag"
+)
+
+type checkTransformer struct {
+	// Config for the entire module.
+	Config *configs.Config
+
+	// Operation is the current operation this node will be part of.
+	Operation walkOperation
+}
+
+var _ GraphTransformer = (*checkTransformer)(nil)
+
+func (t *checkTransformer) Transform(graph *Graph) error {
+	return t.transform(graph, t.Config, graph.Vertices())
+}
+
+func (t *checkTransformer) transform(g *Graph, cfg *configs.Config, allNodes []dag.Vertex) error {
+	moduleAddr := cfg.Path
+
+	for _, check := range cfg.Module.Checks {
+		configAddr := check.Addr().InModule(moduleAddr)
+
+		// We want to create a node for each check block. This node will execute
+		// after anything it references, and will update the checks object
+		// embedded in the plan and/or state.
+
+		log.Printf("[TRACE] checkTransformer: Nodes and edges for %s", configAddr)
+		expand := &nodeExpandCheck{
+			addr:   configAddr,
+			config: check,
+			makeInstance: func(addr addrs.AbsCheck, cfg *configs.Check) dag.Vertex {
+				return &nodeCheckAssert{
+					addr:          addr,
+					config:        cfg,
+					executeChecks: t.ExecuteChecks(),
```

Comment:
Because having nodes in the graph at all can change how the destroy graph is constructed (i.e. these are currently still going to run into problems when placed in an expanding module) , I think it's better to exclude the nodes entirely whenever possible.